### PR TITLE
Menu Items - Right aligned string & Icons for TFT, Fixed lcd_put_u8str_P method.

### PR DESCRIPTION
**Pull request by lukasradek**
### Sumary
Adds ability for generic and static menu items to show (in addition to their title) right aligned string. Visually similar to Edit items but non editable and can be used with ACTION, SUBMENU etc. I have implemented and tested it for DOGM, HD44780 and TFT lcds. I have implemented a simple fallback for e3v2 and TFTGLCD, but I have no way to test it.

Shows helpful menu icons on `TFT_COLOR_UI `in place of `post_char`. Shows the equivalent icons to HD44780 display. `>` is displayed as "right arrow icon", `LCD_STR_UPLEVEL` is displayed as "back icon".

Fixes return value of function `lcd_uint_t lcd_put_u8str_P(PGM_P const ptpl, const int8_t ind, const char *cstr/*=nullptr*/, FSTR_P const fstr/*=nullptr*/, const lcd_uint_t maxlen/*=LCD_WIDTH*/)` in _/lcd/lcdprint.cpp_.
It should have returned number of outputed characters, but was returning number of remaining characters from given `maxlen `limit. Codebase is adjusted to this change, where it was neccessary.

### Detailed Description of Right aligned strings (RString)
**Marlin/src/lcd/menu/menu_item.h**
Adds the option to print RStrings in "any" ITEM. For **full backward compatibility** it is implemented as a companion macro `ITEM_ADD_RIGHT_ALIGNED_STRING`, that user wraps the desired ITEM in. The mechanism of how the string is passed to the `::draw` methods is similar to substituion (more on that later), which requires the macro to also "clear" the string after the item is printed. Otherwise the string would be printed in the following items as well. That fails when screen change is detected (because the menu function immediatelly returns)... therefore there is a small change in `_MENU_INNER_F` needed, that clears the string if action is detected. _Note: It cannot be cleared only after the `ui.screen_changed` is `true`, because when that happens the item action was already performed (that is what caused `ui.screen_changed == true`) and if that action has drawn some items, it would be too late to clear RString._

**Marlin/src/lcd/menu/menu.cpp and .hpp**
RStrings are implemented the same way as the substitution strings are. Supports only C-strings, since I presume it will be mostly used for dynamic values.

**Marlin/src/lcd/tft/ui_common.cpp
Marlin/src/lcd/HD44780/marlinui_HD44780.cpp
Marlin/src/lcd/dogm/marlinui_DOGM.cpp**
RStrings are implemented in `MenuItem_static::draw` and `MenuItemBase::_draw`, which covers most of the _ITEMs. For the others (like EDIT_ITEM) I don't think it is relevant. (It is specified in the `ITEM_ADD_RIGHT_ALIGNED_STRING `macro's comments.
RStrings always have priority over the menu title. That means the whole RString is always visible and has a single space in front of it. Item title then takes the remaining space.
When item's style is `SS_CENTER`, the RString is omited from the padding calculation intentionally.
(There are also a few other changes in these files... see below)

**Marlin/src/lcd/e3v2/marlinui/ui_common.cpp
Marlin/src/lcd/TFTGLCD/marlinui_TFTGLCD.cpp**
Unfortunatelly I have no way to test these LCDs, so I implemented a simple, hopefully error proof fallback solution. RString is just appended to the title.
It is UNTESTED though and should be fully implemented to match the other LCDs.

### Detailed Description of TFT_COLOR_UI icons
**Marlin/src/lcd/tft/ui_common.cpp**
This UI now displays icons in place where `post_chars` would be on HD44780.
It is implemented for `>` character (e.g. submenus), which displays right arrow icon,
and for `up_level` character (e.g. BACK_ITEM), which displays back icon.

### Detailed Description of Other changes
**In generic items, title uses as much width as possible.**
Since `LCD_WIDTH` is a precious real estate, I modified the `MenuItemBase::_draw` so that if `post_char` is only a single space, it is ignored and item title will use that character. The only "downside" is that submenu symbol `>` can be shown at the same column that other items' titles can be, making the distinction between submenu and other items a little less noticeable. But in most places this makes no difference (most item titles are not that long), but when there is a long title, that one more character can help.
This concerns these files:
Marlin/src/lcd/tft/ui_common.cpp
Marlin/src/lcd/HD44780/marlinui_HD44780.cpp
Marlin/src/lcd/dogm/marlinui_DOGM.cpp

**Return value fix for `lcd_uint_t lcd_put_u8str_P(PGM_P const ptpl, const int8_t ind, const char *cstr/*=nullptr*/, FSTR_P const fstr/*=nullptr*/, const lcd_uint_t maxlen/*=LCD_WIDTH*/)`**
This is the function that handles item title substitution. Documentation (of this function and some of its callers) says that it should `@return the output width (in pixels on GLCD)`, but it did not. I returned the width of unused space of the `maxlen` parameter limit.
I have also manually searched through the codebase, made sure that this change did not break anything and fixed places where it did (mostly `::draw` methods that I was modifying for the other features).
Implementation change concerns this file:
Marlin/src/lcd/lcdprint.cpp
Touch ups where return value was used concerns these:
Marlin/src/lcd/tft/ui_common.cpp
Marlin/src/lcd/HD44780/marlinui_HD44780.cpp
Marlin/src/lcd/dogm/marlinui_DOGM.cpp
Marlin/src/lcd/e3v2/marlinui/ui_common.cpp
Marlin/src/lcd/TFTGLCD/marlinui_TFTGLCD.cpp

### Benefits

**Right aligned strings in Item titles**
are needed for my next pull request regarding reworked (Assisted) Tramming Wizard.
It can be useful in general to display separate non editable value inside menu item title, while not being restricted to EDIT_ITEM's properties (cannot open menus, cannot trigger actions).

**TFT_COLOR_UI icons**
It is nicer looking 🙂. It is also very practical being able to distinguish between SUBMENU and ACTION items.

**Other**
Collapsing `post_char` can help on smaller screens.
Fixed return value makes more sense this way and is consistent with other functions.

### Related Issues

It is a first part of completing [this feature request](https://github.com/MarlinFirmware/Marlin/issues/24448)

### Thousands of words
RStrings and post_char icons working side by side for ACTION items and submenus alike in my [new Tramming Wizard](https://github.com/MarlinFirmware/Marlin/pull/25339).
<p float="left">
  

</p>

### CODE REVIEW

**Comment by lukasradek** [comment]
Force-pushes were just a slight refactor to improve code readability, guard agains unreasonably long RStrings and UI visual touch up.
No more coming, this is it 🙂.

**Comment by jmz52** [comment]
There is a method to display right-aligned text in menu items [ui_common.cpp#L154](https://github.com/MarlinFirmware/Marlin/blob/2.1.2/Marlin/src/lcd/tft/ui_common.cpp#L154)
Can it be used instead of creating a itemRAlignedStringC entity that affects all menu items and needs to be manually reset after each use?

**Comment by lukasradek** [comment]
TLDR: 
_"There is a method to display right-aligned text in menu items."_ I will correct that statement a little.
It is a method to display a specific EDIT_ITEM with right aligned string. It does not display right aligned string in MENU_ITEMs in general.
I know this looks like a big intervention in the code (and raises the "potential trouble" red flag), but it is not, really. It is not a hot shot, I was on it for about a week, trying to figure out the least intrusive solution. I considered it, thought it through (before implementing this) and nope, using `MenuEditItemBase::draw` it is not possible / practical.

-----

**Full reasoning:**
I needed to have items with right aligned value (RString), that would **trigger an action**, act as a **submenu** and/or just sit there (**static item**).

**If I was to use `MenuEditItemBase:draw`, I would have to go in one of those ways...**

1. **Try to use `EDIT_ITEM` macro.** That would allow me to display right aligned value, but it would not fulfill any of the remaining requirements. Trying to extend `EDIT_ITEM` to add those features wouldn't be nice nor semantic.
2. **Creating my own `_ITEM` macros that would use `MenuEditItemBase:draw`.** I guess that would work in the end but I would have to create macros for `STATIC_ITEM_RSTRING`, `ACTION_ITEM_RSTRING `and `SUBMENU_ITEM_RSTRING`... and if I wanted to make this extension flexible and not just for myself, I would have to create all those `_F`, `_S_F`, `_N_F`, `_N_S_F`, ... variants, which would be huuuuge clutter and code duplication (also on the drawing side).
I would have to mimic the behaviour of the original `_submenu::draw`, `_static::draw` and `_function::draw` which would again lead to duplication and loss of single source of truth (if anybody wanted to change submenu `pre_char`, he would have to modify the original `_submenu::draw` and my mimicking draw method as well).
Furthermore `MenuEditItemBase::draw` does not handle the situation where `"left string" + "right string" > LCD_WIDTH` (at least in the TFT version you have linked).

And so I decided to implement `ITEM_ADD_RIGHT_ALIGNED_STRING` macro because:
1. **Full backward compatibility.** "It affects all menu items" (as you said)... but it doesn't actually. Developers can still use all `_ITEM` macros as before... nothing changes. On the `::draw` side, it is implemented so that the overhead is minimal. If `itemRAlignStringC == nullptr` (which it is everytime unless you **want** the RString), the only overhead is just adding/subtracting `0` once or twice (or do nothing in TFT code).
2. **Preventing code duplication, while preserving all the functionality.** Developer can use "any" of the already existing pallete of `_ITEM` macros with their specific functionality. And optionally wrap it in my macro to use RString. There is no code duplication anywhere (neither the macro side nor `::draw` side).
4. **Ease of use.** Developer wants right aligned string? Just wrap your already existing `MENU_ITEM` macro in my `ITEM_ADD_RIGHT_ALIGNED_STRING`. Done!
**There is NO need to reset it manually after each use** as you have written. The macro takes care of it by itself seamlessly for the developer.
_(Btw... a little side note... substitution strings `itemStringC / F` are implemented the exact same way and nobody bothers to reset them. They only rely on the fact that noone will use substitution placeholder in item title without actually setting their own substituion string. Try it... if you use substitution (some `ITEM_S`) macro and then use non-substitution macro BUT with a string including "$", it will substitute the previous item's string anyway. Similar unintended substitution applies to other placeholders as well... `=` for example.)_
```
SUBMENU_S_F("Hello, still here",F("Hi. $"), []{});   // Resulting item: Hi. Hello, still there
SUBMENU_F(F("Goodbye. $"), []{});                    // Resulting item: Goodbye. Hello, still there
```
5. **Robustness**. The `::draw` methods (that I modified for RStrings) can now gracefully handle situations, where the two string would not fit on the screen in whole or even when any of those is wider than the screen itself.

**Another nice use case outside my Tramming Wizard**
Right now language menu is shown just as a button with current language name.
`English                  >` //  Yes.. it is functional.
But it would be nice to also make Marlin a bit more user friendly and polished.
With RStrings, we can quite easily make the menu item looks like
`Language:        English >`
while still being a submenu.

**Comment by lukasradek** [comment]
Ignore that last force push from d0c1cc2 to c7c496c. My mistake. Sorry.
There are no changes - c7c496c is still the relevant commit to this PR.

**Comment by jmz52** [comment]
I see. 
Would you mind making the display of `post_char` optional by putting corresponding code behind somethings like
`#ifdef TFT_UI_SHOW_CLASSIC_MARKS` or `#ifndef TFT_UI_HIDE_CLASSIC_MARKS` ?
I personally prefer the UI without these marks.

**Comment by lukasradek** [comment]
Really? 🙂 And how do you tell ACTION from SUBMENU without these 🙂.
I sure can do that... one question though. Who is the authority to decide how much configurable things should be?
Even though I guess the merger can cherry pick it out if neccessary.

**Comment by jmz52** [comment]
Somehow this never was a problem.
Either names are descriptive enough or I just spent too much time looking at them from the inside. 🙂
As I remember there was no feature request to add these marks for more than two years now, so probably I am not the only one who is fine with UI without them.
It is up to you whether to make it configurable , I'll surely have no problems to comment out few line for my personal build.

From what I know the only authority is Marlin admin team.
"Regular users can open and close their own issues, but only the administrators can do project-related things like add labels, merge changes, set milestones, and kick trolls."
